text/0000-iri-v181-transaction-and-bundle/0000-iri-v181-transaction-and-bundle.md

```diff
@@ -0,0 +1,980 @@
+    /// After validation, transactions of a bundle are candidates to be included to the ledger.
+    /// 
+    /// For a bundle to be considered valid, the following assertions must be true:
+    /// 
+    /// + bundle has the correct length;
+    /// + transactions share the same bundle hash;
+    /// + transactions absolute value doesn't exceed total IOTA supply;
+    /// + bundle absolute sum never exceeds total IOTA supply;
+    /// + order of transactions in the bundle is the same as announced by `current_index` and `last_index`;
+    /// + input/output transactions have an address ending in `0` i.e. has been generated with Kerl;
+    /// + bundle inputs and outputs are balanced i.e. the bundle sum equals `0`;
+    /// + announced bundle hash matches the computed bundle hash;
+    /// + for input transactions, the signature is valid;
+    /// 
+    /// TODO: `IOTA_SUPPLY` used below is some global constant which needs to be defined and set externally.
+    pub fn validate<S: Sponge>(&mut self, sponge: S) -> Result<Self, OutgoingBundleError> {
+        unimplemented!()
+
+        let mut value = 0;
+        let mut current_index = 0;
+
+        // NOTE: We are not defining `first_transaction` here but leave it to the implementation phase.
+        if bundle_builder.len() != bundle_builder.first_transaction().last_index + 1 {
+            Err(InvalidLength)?
+        }
+
+        let bundle_hash = bundle_builder.first_transaction().bundle_hash;
+        let last_index = bundle_builder.first_transaction().last_index;
+
+        for transaction_draft in bundle_builder {
+            if transaction_draft.bundle_hash != bundle_hash {
+                Err(InvalidHash)?
+            }
+
+            if abs(transaction_draft.value) > IOTA_SUPPLY {
+                Err(InvalidTransactionValue)?
+            }
+
+            value += transaction_draft.value;
+            if abs(value) > IOTA_SUPPLY {
+                Err(InvalidValue)?
+            }
+
+            if transaction_draft.current_index != current_index++ {
+                Err(InvalidIndex)?
+            }
+
+            if transaction_draft.last_index != last_index {
+                Err(InvalidIndex)?
+            }
+
+            if transaction_draft.value != 0 && transaction_draft.address.last != 0 {
+                Err(InvalidAddress)?
+        }
+
+        if value != 0 {
+            Err(InvalidValue)?
+        }
+
+        if bundle_hash != self.calculate_hash() {
+            Err(InvalidHash)?
+        }
+
+        // NOTE: We do not provide an example of `validate_withdrawal_transactions`. Its purpose is to walk
+        //       through those transactions in the bundle builder and verify their signatures.
+        bundle_builder.validate_withdrawal_transactions()?;
+
+        Ok(Self)
+    }
+
+    pub fn build(self) -> Result<Bundle, OutgoingBundleError> {
```

Comment:
In the end, `validate` and `build` from `IncomingBundleBuilder ` and `SealedOutgoingBundleBuilder ` should be the same. What can we do for it ?

Reply:
We can probably implement one in terms of the other.